modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/odata.es.js

```diff
@@ -0,0 +1,345 @@
+ * @param {object} oDataASTNode
+ * @returns String value of the internal name.
+ */
+function getChildExpressionName(oDataASTNode) {
+	return getExpressionName(oDataASTNode.value);
+}
+
+/**
+ * Gets the conjunction of the group or returns AND as a default.
+ * @param {object} oDataASTNode
+ * @returns The conjunction name for a group or, if not available, AND.
+ */
+function getConjunctionForGroup(oDataASTNode) {
+	const childExpressionName = getChildExpressionName(oDataASTNode);
+
+	return isValueType(CONJUNCTIONS, childExpressionName) ? childExpressionName : CONJUNCTIONS.AND;
+}
+
+/**
+ * Gets the internal name of an expression from the oDataV4Parser name.
+ * @param {object} oDataASTNode
+ * @returns String value of the internal name
+ */
+function getExpressionName(oDataASTNode) {
+	return oDataV4ParserNameMap[oDataASTNode.type];
+}
+
+/**
+ * Returns the next expression in the syntax tree that is not a grouping.
+ * @param {object} oDataASTNode
+ * @returns String value of the internal name of the next expression.
+ */
+const getNextNonGroupExpressionName = oDataASTNode => {
+	let returnValue;
+
+	if (oDataASTNode.value.type === 'BoolParentExpression') {
+		returnValue = getNextNonGroupExpressionName(oDataASTNode.value);
+	}
+	else {
+		returnValue = oDataASTNode.value.left ?
+			oDataASTNode.value.left :
+			oDataASTNode.value;
+	}
+
+	return getExpressionName(returnValue);
+};
+
+/**
+ * Checks if a grouping has different conjunctions (e.g. (x AND y OR z)).
+ * @param {boolean} lastNodeWasGroup
+ * @param {object} oDataASTNode
+ * @param {string} prevConjunction
+ * @returns boolean of whether a grouping has different conjunctions.
+ */
+function hasDifferentConjunctions({lastNodeWasGroup, oDataASTNode, prevConjunction}) {
+	return prevConjunction !== oDataASTNode.type && !lastNodeWasGroup;
+}
+
+/**
+ * Checks if the criteria is a group by checking if it has an `items` property.
+ * @param {object} criteria
+ */
+function isCriteriaGroup(criteria) {
+	return !!criteria.items;
+}
+
+/**
+ * Checks if the value is a certain type.
+ * @param {object} types A map of supported types.
+ * @param {*} value The value to validate.
+ * @returns {boolean}
+ */
+function isValueType(types, value) {
+	return Object.values(types).includes(value);
+}
+
+/**
+ * Checks if the group is needed; It is unnecessary when there are multiple
+ * groupings in a row, when the conjunction directly outside the group is the
+ * same as the one inside or there is no conjunction within a grouping.
+ * @param {boolean} lastNodeWasGroup
+ * @param {object} oDataASTNode
+ * @param {string} prevConjunction
+ * @returns a boolean of whether a group is necessary.
+ */
+function isRedundantGroup({lastNodeWasGroup, oDataASTNode, prevConjunction}) {
+	const nextNodeExpressionName = getNextNonGroupExpressionName(oDataASTNode);
+
+	return lastNodeWasGroup || oDataV4ParserNameMap[prevConjunction] === nextNodeExpressionName || !isValueType(CONJUNCTIONS, nextNodeExpressionName);
```

Comment:
You might want to add some `()` here since it is pretty confusing to read. I'm not totally sure what order things evaluate.

Reply:
I'll add some new lines in here to divide each check. It's a flat list of `||` so adding parenthesis shouldn't matter.
```javascript
	return lastNodeWasGroup ||
		oDataV4ParserNameMap[prevConjunction] === nextNodeExpressionName ||
		!isValueType(CONJUNCTIONS, nextNodeExpressionName);
```

Reply:
Oh great, that looks much better.